imports/lib/models/SoftDeletedModel.ts

```diff
@@ -0,0 +1,200 @@
+import { Mongo } from 'meteor/mongo';
+import { z } from 'zod';
+import { deleted } from '../schemas/customTypes';
+import type { ValidTopLevelType } from '../schemas/generateJsonSchema';
+import type { ModelType, Selector, SelectorToResultType } from './Model';
+import Model from './Model';
+
+const injectQuery = <S extends Mongo.Selector<any>>(
+  selector: S | string | Mongo.ObjectID | undefined,
+  injection: S,
+) => {
+  if (!selector) {
+    return injection;
+  }
+  if (typeof selector === 'string') {
+    return { _id: selector, ...injection };
+  }
+  if (selector instanceof Mongo.ObjectID) {
+    return { _id: selector, ...injection };
+  }
+  if ('$query' in selector) {
+    return {
+      ...selector,
+      $query: {
+        $and: [
+          selector.$query,
+          injection,
+        ],
+      },
+    };
+  }
+
+  return {
+    $and: [
+      selector,
+      injection,
+    ],
+  };
+};
+
+const injectOptions = <Opts extends Mongo.Options<any>>(
+  options: Opts | undefined,
+) => {
+  if (options?.fields) {
+    return {
+      ...options,
+      fields: {
+        ...options.fields,
+        deleted: 1,
+      },
+    };
+  }
+
+  return options;
+};
+
+class SoftDeletedModel<Schema extends ValidTopLevelType>
+  extends Model<
+    Schema extends z.ZodObject<infer Shape, infer UnknownKeys, infer Catchall> ?
+      z.ZodObject<Shape & { deleted: typeof deleted }, UnknownKeys, Catchall> :
+      z.ZodIntersection<Schema, z.ZodObject<{ deleted: typeof deleted }>>
+  > {
+  constructor(name: string, schema: Schema) {
+    super(
+      name,
+      schema instanceof z.ZodObject ?
+        schema.merge(z.object({ deleted })) :
+        schema.and(z.object({ deleted })) as any
+    );
+  }
```

Comment:
This ends up causing a bit of a style change in how we export types for our models.

I tried to write this class so that the schema being passed in had to already have a `deleted` property on it, but I couldn't figure out how to express that as a type constraint on the `Schema` type, so instead I decided to add the `deleted` property as part of the construction of a `SoftDeletedModel`.

That in turn means that the schema you pass in isn't necessarily reflective of the actual type of the collection. Therefore, instead of exporting the record type from the schema file, I added a new `ModelType` helper, with the expectation that you export the type in the model declaration, not the schema declaration.

This will churn imports everywhere a little bit (look around for e.g. `SettingType`), but I couldn't figure out any other way to do it. I think the minor upside is that it lets us potentially write more powerful extensions of model functionality if we want to.

Reply:
Seems fine.  This also has the potential upside of making app code only import from `models`, instead of importing the collection from `models` and the type from `schemas`, which feels like maybe a slightly cleaner layering?